tests/per_provider/normal_tests/import_key.rs

```diff
@@ -69,25 +82,162 @@ fn import_key_twice() -> Result<()> {
+    let mut client = TestClient::new();
+    let key_name = String::from("check_format_import3");
+
+    let public_key = RsaPublicKey {
+        modulus: IntegerAsn1::from_unsigned_bytes_be(vec![0xDE; 1024]),
+        public_exponent: IntegerAsn1::from_unsigned_bytes_be(vec![0x01, 0x00, 0x01]),
+    };
+
+    let attributes = KeyAttributes {
+        key_type: KeyType::RsaPublicKey,
+        key_bits: 1023,
+        key_policy: KeyPolicy {
+            key_usage_flags: UsageFlags {
+                sign_hash: false,
+                verify_hash: true,
+                sign_message: false,
+                verify_message: true,
+                export: false,
+                encrypt: false,
+                decrypt: false,
+                cache: false,
+                copy: false,
+                derive: false,
+            },
+            key_algorithm: Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign {
+                hash_alg: Hash::Sha256,
+            }),
+        },
+    };
+
     let status = client
         .import_key(
             key_name,
-            KeyType::RsaPublicKey,
-            Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign {
+            attributes,
+            picky_asn1_der::to_vec(&public_key).unwrap(),
+        )
+        .unwrap_err();
+
+    assert_eq!(status, ResponseStatus::PsaErrorInvalidArgument);
+
+    Ok(())
+}
+
+#[test]
+fn failed_imported_key_should_be_removed() -> Result<()> {
+    // If the key_bits field of the key attributes is zero, the operation should still work.
+    // The size of the key is always taken from the data parameter.
+    let mut client = TestClient::new();
+    let key_name = String::from("failed_imported_key_should_be_removed");
+
+    let public_key = RsaPublicKey {
+        modulus: IntegerAsn1::from_unsigned_bytes_be(example_modulus_1024()),
+        public_exponent: IntegerAsn1::from_unsigned_bytes_be(vec![0x01, 0x00, 0x01]),
+    };
+
+    let attributes = KeyAttributes {
+        // Not supported
+        key_type: KeyType::Aes,
+        key_bits: 1024,
+        key_policy: KeyPolicy {
+            key_usage_flags: UsageFlags {
+                sign_hash: false,
+                verify_hash: true,
+                sign_message: false,
+                verify_message: true,
+                export: false,
+                encrypt: false,
+                decrypt: false,
+                cache: false,
+                copy: false,
+                derive: false,
+            },
+            key_algorithm: Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign {
                 hash_alg: Hash::Sha256,
             }),
-            KEY_DATA.to_vec(),
-        )
-        .expect_err("The key with the same name has already been created.");
-    assert_eq!(status, ResponseStatus::PsaErrorAlreadyExists);
+        },
+    };
+
+    if client
+        .import_key(key_name.clone(), attributes, Vec::new())
+        .is_err()
```

Comment:
This is kinda wrong - if `import_key` isn't an error, the test still succeeds. You should probably `?` it and just carry on with the next import.

Reply:
There are others like this - `failed_created_key_should_be_removed` is one example.

Reply:
I did it like this because the way I trigger an error here is by giving a key type of `Aes` which is not currently supported. However, we might not want this test to fail the moment `Aes` is supported, or does not trigger an error.

Reply:
> However, we might not want this test to fail the moment Aes is supported

The import will fail when AES is supported anyway, because AES keys have at most 256 bits

> or does not trigger an error.

I'd say it should fail, because otherwise what's the point of the test? You want it to fail on the first import to show that you can import again succesfully after - imagine if we do something which leads to the first import succeeding, but now if you fail an import you can't reuse the name - we wouldn't be testing for that anymore so we wouldn't know.

Reply:
> The import will fail when AES is supported anyway, because AES keys have at most 256 bits

Good point!
I think I confused myself anyway: this test works in two steps:
1. We **intentionally** make an operation that will fail when generating/importing a key.
2. We try to generate/import a key with the same name to check that the key from 1 has been destroyed.

If I used `?` instead of `is_err`, on the first import, then the test will fail without us actually testing anything 😋

Reply:
> If I used ? instead of is_err, on the first import, then the test will fail without us actually testing anything 😋

Reply:
Yes good point, so we also check that the first one is definitely an error.